### DISCUSSION TITLE
per class AP, and metrics

### DISCUSSION THREAD

**Discussion by benlten** [ultralytics/yolov5#2605] [Q&A]
Unless I am misreading the code return, the function: 
ap_per_class in (https://github.com/ultralytics/yolov5/blob/master/utils/metrics.py)

returns the mean AP and not a list of APs. 

**Goal**
I am trying to plot how each individual class is doing, or rather the precision, AP50,  and recall per class label as would be shown in a confusion matrix.  

Currently I see wandb show the average per P, R, AP50 etc - but I'd like to log a dict of those per individual class performance to analyze the misclassification errors. 

If someone has gone through the same steps to save data out, please let me know.

**Comment by glenn-jocher** [comment]
@bencten confusion matrices and per-class statistics are automatically reported by test.py on final epoch, or when called directly (i.e. python test.py):

See https://github.com/ultralytics/yolov5/pull/1474 for confusion matrix:
![image](https://user-images.githubusercontent.com/26833433/112555195-ff875680-8dc7-11eb-8948-e9cb15b183ea.png)

test.py verbose (per class) output is enabled automatically for small datasets, or can be forced manually with `python test.py --verbose`. See https://github.com/ultralytics/yolov5/pull/1997
![image](https://user-images.githubusercontent.com/26833433/112557389-a837b500-8dcc-11eb-80a1-c7dbd51802c7.jpg)

**Reply by benlten** [reply]
@glenn-jocher 
There may be a bug. I was tracing through the code to the metrics method for a while. 
In my csv dump, I kept seeing absurdly high numbers as if something was being cached coming from when test.py returned. 
**class_name 6 1 1.0 1.0 0.9954166666666666 0.9954166666666667**, no matter what. 
I'll keep an eye on this and present a small reproducible test set if it occurs again.

**Reply by glenn-jocher** [reply]
@bencten thanks for the info! If you spot any reproducible bugs, please raise a new issue using the 🐛 **Bug Report template** with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. Thank you!
